README.md
# FlightGearBuilder
## About
#### FlightGearBuilder is an elegant and simple script that will:
- On Debian based distros, it will download and install all depedencies required to build FlightGear and its components from the source.
- Download the source code of PLIB, OpenSceneGraph, SimgGear and, FlightGear.
- Compile each of the components after its source is downloaded.

## What Does It Do?
- **FlightGearBuilder** is a shell script that will download the source code for OSG, PLIB, SimGear and FlightGear.
- Then it will compile a nice and stable version of them. Or unstable, if so you chose.

## How do I run it? How to use it?
#### Installing FlightGear
1. Download the script run it with **`./FlightGearBuilder.sh`**. You may or may not need to mark it as executable first with **`chmod +x FlightGearBuilder.sh`**
- Add **`--next`** flag to each script if you want to download the very
latest testing version of FlightGear.

## FGDataDownloader.sh
## About FGDataDownloader.sh
- **GDataDownloader** is a tiny script that will download the "Data" required to run FlightGear.
- The Data will be downloaded to the FGB folder, tipically located in your home folder.

## Running FGDataDownloader.sh
1. Download the script run it with **`./FGDataDownloader.sh`**. You may or may not need to mark it as executable first with **`chmod +x FGDataDownloader.sh`**
- Add **`--next`** flag to each script if you want to download the very
latest testing version of FlightGear.

# FlightGear Data
- FlightGear Data are required files to run FlightGear.
- https://download.flightgear.org/builds/
- https://gitlab.com/flightgear/fgdata

## Uninstalling FlightGear
1. Simply remove the **`FGB`** directory from your **`$HOME`** folder.
2. There will be a launcher at **`$HOME/.local/share/applications/`**, remove it to remove FlightGear from the applications menu.

## Where is FlightGear Installed to?
- By default the script will install to **`~/FGB`**.

## How to I run FlightGear?
- If everything wel well, you can run **`./flightgear --launcher`** from the folder **`~/FGB/FlightGear-Stable/Next`**.
- Or, by simply searching for and clicking in **FlightGear** in your DEs applications menu/dashboard.


CHANGELOG
# CHANGELOG

# Nue Generation
# 2.20210816.Beta1
# Work done on puting variable names in quotes.
# Minor text changes throughout.
# Compiler flags defined on the top of the code in a single place now.
# Install "root" now defined by a properly named variable. Making easier to
# move FG to another location.
# Naming convention will be as the following: MajorRelease.MinorRelease.StabilityState
# Updated readme.

# 2.20210521.Alpha1
# Totally new release, rewritten from scratch, now using Shell functions.
# Added ability to download dependencies automatically on Debian based distros.
# Removed ability to donwload FlightGear data.
# Changed versioning scheme to MajorVersion.MinorVersion.StabilityState

# v2.0-Beta-1
# Shortcuts are added to the Desktop and the desktop applications menu.

# v2.0-Beta
# New major release
# flightgear_downloader renamed to FlightGear_Downloader.sh
# flightgear_compiler renamed to FlightGear_Compiler.sh
# fgdata_downloader renamed to Download_Data.sh
# Major changes on the functioning of the script.
# By default FlightGear is installed to $HOME/FlightGear-Stable,
# the next version, selected by using --next, installed to $HOME/FlightGear-Next
# Uninstaller removed, to uninstall simply delete the directory.
# Due to FlightGear devs mistake, now stable version is built against the 
# release/2020.3 branch instead of version/2020.3.6 tag.

# v1.1.2020.12.23.02.14
# Update downloader to download latest stable version of FlightGear, 2020.3.5

# v1.1.2020.12.14.02.42
# Fix scenery.

# v1.1.2020.12.14.23.25
# Just some house keeping.
# Completed setting all FG environment variables.
# Done a bit more variables safeguarding.

# v1.1.2020.12.09.13.17
# Fixed many variable attributions.
# Walking towards POSIX compliance.
# Fixed many typos.
# Added a nice fgdata downloader, with options to download stable and testing versions of it.
# More code tweaks.
# Version bump!

# v1.0.2020.12.05
# Added an uninstaller.
# Code tweaks.
# First stable version.
# Tells user to press ctrl c to quit the script.
# Updated README.md to reflect changes done to this version.

# v1.0.2020.12.04
# build cpu type set to native, O flag to O3
# tempdir variable renamed to download_dir
# User now able to select next/testing version instead of stable.
# Changelog moved to CHANGELOG file.
# builderversion renamed to version and moved to VERSION file.
# Small change in version numbering.
# First stable version.
# Script file renamed to flightgear_builder.
# Script split in two parts, flightgear_builder which will compile FlightGear from source and flightgear_downloader which will only download and update the source.

# v0.8.20201204
# Added more code comments.
# Changed EOF section to add fgfs runner in one step.

# v0.7.20201202
# OSG broken with GLNV, reverting back to LEGACY.
# Fix some typos.
# Removing script to download FGData. Alternative will be given in the text interface.
# Including release date with release version.
# Now some clean up is done prior running the script, for cleaner builds.

# v0.6
# Fix bad bad bad typos and variables declarations to flightgear runner.
# Added Spaces and empty lines to CLI.
# Defaulting back to release/2020.3.
# GLVND driver is default now, before it was LEGACY.

# v0.5
# Switched OSG to OSGs official GIT repo and to branch 3.6.
# Compiling flags are now more conservative.
# Minor text tweaks and code changes.

# v0.4
# No longer downloads FlightGear data. DownloadFGdata.sh should be used for that instead.

# v0.3
# No longer creates launcher at .local/bin and no longer changes PATH.
# Moved some OSD git URL to variable.
# Moved nproc to variable.


Download_Data.sh
#!/bin/bash

### ABOUT
# FlightGear_Data_Downloader.sh
# This script will download, from Git repositories, the data files
# required to run the FlightGear Flight Simulator.

### AUTHOR
# This software was created by Megaf - https://github.com/Megaf
# On the 20/10/2020

### LICENSE
########################################################################
#                                                                      #
#                       FREE AS IN FREEDOM LICENSE                     #
#                           FAIFL Version 0.2                          #
#                             October 2020                             #
#                                                                      #
#           Copyright (C) 2020 Megaf (https://github.com/Megaf/)       #
#                     <mmegaf [at] gmail [dot] com>                    #
#                                                                      #
# Everyone, anyone, anything or any company or organization, human,    #
# plant, apache helicopter, alive or not, form of life or not, is/are  #
# allowed to copy, modify, distribute, change, charge or do anything   #
# they/ones want with anything or whatever this license is, attached   #
# to or licensing.                                                     #
#                                                                      #
# This thing, software, hardware, piece, object, art, or whatever that #
# is been licensed under FAIFL (Free As In Freedom License) can be     #
# used, shared, copied, sold, modified, changed, adapted, or whatever  #
# you want to do as long as the following terms are understood and met.#
#                                                                      #
# (1) This that is licensed under FAIFL comes as it is.                #
# (2) This that is licensed under FAIFL it is not covered under any    #
# warranty.                                                            #
# (3) There is no warranty or guarantee that this that is licensed     #
# under FAIFL works at all.                                            #
# (4) This that is licensed under FAIFL may be dangerous and cause     #
# harm and loss of property.                                           #
# (5) Changing this license is not allowed.                            #
# (6) The creator, owner, author of this that is licensed under FAIFL  #
# shall not be liable,responsible or accused in any way for anything   #
# at all.                                                              #
#                                                                      #
########################################################################

# Using read command to get version information from the file VERSION
read -r version < VERSION

# Directory where the source files will be downloaded to.
install_directory="$HOME/FlightGear-Stable"

# Git Repository to download the source files from.
fgdata_repository="git://git.code.sf.net/p/flightgear/fgdata"

# Branches/Versions of each component.
#fgdata_branch="version/2020.3.6"
fgdata_branch="release/2020.3"

# By default, without command line arguments the script will download the stable
# version of FlightGear, the following "if" statement downloads the "next"
# development version of FlightGear.
#
# It will also install FlightGear to folder name FlightGear-Next instead of 
# FlightGear-Stable
if [ "$*" = "--next" ]; then
    install_directory="$HOME/FlightGear-Next"
    fgdata_branch="next"
fi

### INTRODUCTION PART
clear
echo "#====== Welcome to FlightGearBuilder $version !"
echo "#====== This script will download FlightGear data files."
echo ""
echo "#====== Use --next to download the development version of FGData."
echo ""

# Waits for the user to press a key to continue.
echo ""
echo "#====== Press any key to continue or Ctrl + C to cancel."
read -rsn1
clear

# Checking FlightGear Data.
echo ""
echo "#====== Checking if FlightGear Data was already downloaded."
if [ -d "$install_directory"/data ]; then
    echo "#======  FlightGear Data found, updating the repo if neded."
    cd "$install_directory"/data && git pull # Try to update existing previously cloned repo.
        else
            echo "#====== FlightGear Data not found, downloading now."
            git clone -b "$fgdata_branch" "$fgdata_repository" "$install_directory"/data # Clone repo from Git
fi


DownloadFGdata.sh
#!/bin/bash

# DownloadFGdata.sh v0.2

### AUTHOR
# This software was created by Megaf - https://github.com/Megaf
# On the 24/10/2020

### LICENSE
########################################################################
#                                                                      #
#                       FREE AS IN FREEDOM LICENSE                     #
#                           FAIFL Version 0.2                          #
#                             October 2020                             #
#                                                                      #
#           Copyright (C) 2020 Megaf (https://github.com/Megaf/)       #
#                     <mmegaf [at] gmail [dot] com>                    #
#                                                                      #
# Everyone, anyone, anything or any company or organization, human,    #
# plant, apache helicopter, alive or not, form of life or not, is/are  #
# allowed to copy, modify, distribute, change, charge or do anything   #
# they/ones want with anything or whatever this license is, attached   #
# to or licensing.                                                     #
#                                                                      #
# This thing, software, hardware, piece, object, art, or whatever that #
# is been licensed under FAIFL (Free As In Freedom License) can be     #
# used, shared, copied, sold, modified, changed, adapted, or whatever  #
# you want to do as long as the following terms are understood and met.#
#                                                                      #
# (1) This that is licensed under FAIFL comes as it is.                #
# (2) This that is licensed under FAIFL it is not covered under any    #
# warranty.                                                            #
# (3) There is no warranty or guarantee that this that is licensed     #
# under FAIFL works at all.                                            #
# (4) This that is licensed under FAIFL may be dangerous and cause     #
# harm and loss of property.                                           #
# (5) Changing this license is not allowed.                            #
# (6) The creator, owner, author of this that is licensed under FAIFL  #
# shall not be liable,responsible or accused in any way for anything   #
# at all.                                                              #
#                                                                      #
########################################################################

### ABOUT THIS SOFTWARE
# This script will download, from Git repositories, the FGDATA files.

# Directory where everything will be installed at.
installdir="$HOME/FlightGear"

# Variables defining which branches will be used for each repository.
fgbranch="next" # Branches for FlightHear and SimGear

### CHECKS AND GIT CLONE
# We will perform a series of checks to see if the repos were cloned
# already or not, if they were not cloned yet, then git clone will run,
# Otherwise, the script will try to update, git pull, the repos.

# Checking FlightGear Data.
echo "#====== Checking if FlightGear data was already downloaded."
if [ -d "$installdir/data" ]; then
echo "#======  FlightGear Data found, updating the repo if neded."
cd $installdir/data && git pull
else
echo "#====== FlightGear Data not found, downloading now."
git clone --depth=1 -b $fgbranch git://git.code.sf.net/p/flightgear/fgdata $installdir/data
fi


FGDataDownloader.sh
#!/bin/bash

### ABOUT
# FlightGear_Data_Downloader.sh
# This script will download, from Git repositories, the data files
# required to run the FlightGear Flight Simulator.

### AUTHOR
# This software was created by Megaf - https://github.com/Megaf
# On the 20/10/2020

### LICENSE
########################################################################
#                                                                      #
#                       FREE AS IN FREEDOM LICENSE                     #
#                           FAIFL Version 0.2                          #
#                             October 2020                             #
#                                                                      #
#           Copyright (C) 2020 Megaf (https://github.com/Megaf/)       #
#                     <mmegaf [at] gmail [dot] com>                    #
#                                                                      #
# Everyone, anyone, anything or any company or organization, human,    #
# plant, apache helicopter, alive or not, form of life or not, is/are  #
# allowed to copy, modify, distribute, change, charge or do anything   #
# they/ones want with anything or whatever this license is, attached   #
# to or licensing.                                                     #
#                                                                      #
# This thing, software, hardware, piece, object, art, or whatever that #
# is been licensed under FAIFL (Free As In Freedom License) can be     #
# used, shared, copied, sold, modified, changed, adapted, or whatever  #
# you want to do as long as the following terms are understood and met.#
#                                                                      #
# (1) This that is licensed under FAIFL comes as it is.                #
# (2) This that is licensed under FAIFL it is not covered under any    #
# warranty.                                                            #
# (3) There is no warranty or guarantee that this that is licensed     #
# under FAIFL works at all.                                            #
# (4) This that is licensed under FAIFL may be dangerous and cause     #
# harm and loss of property.                                           #
# (5) Changing this license is not allowed.                            #
# (6) The creator, owner, author of this that is licensed under FAIFL  #
# shall not be liable,responsible or accused in any way for anything   #
# at all.                                                              #
#                                                                      #
########################################################################

# Using read command to get version information from the file VERSION
read -r version < VERSION
instroot="$(envsubst < DESTINATION)"

# Directory where the source files will be downloaded to.
install_directory="$instroot""/FGB/FGData-Stable"

# Git Repository to download the source files from.
fgdata_repository="git://git.code.sf.net/p/flightgear/fgdata"
#fgdata_repository="https://gitlab.com/flightgear/fgdata"

# Branches/Versions of each component.
fgdata_branch="release/2020.3"

# By default, without command line arguments the script will download the stable
# version of FlightGear, the following "if" statement downloads the "next"
# development version of FlightGear.
#
# It will also install FlightGear to folder name FlightGear-Next instead of
# FlightGear-Stable
if [ "$*" = "--next" ]; then
    install_directory="$instroot""/FGB/FGData-Next"
    fgdata_branch="next"
fi

### INTRODUCTION PART
clear
echo "#====== Welcome to FlightGearBuilder $version !"
echo "#====== This script will download FlightGear data files."
echo ""
echo "#====== Use --next to download the development version of FGData."
echo ""

# Waits for the user to press a key to continue.
echo ""
echo "#====== Will get data $fgdata_branch - Press any key to continue or Ctrl + C to cancel."
#read -rsn1
clear

# Checking FlightGear Data.
echo ""
echo "#====== Checking if FlightGear Data was already downloaded."
if [ -d "$install_directory"/data ]; then
    echo "#======  FlightGear Data found, updating the repo if neded."
    cd "$install_directory"/data && git pull # Try to update existing previously cloned repo.
        else
            echo "#====== FlightGear Data not found, downloading now."
            mkdir -p "$install_directory"
            git clone -b "$fgdata_branch" "$fgdata_repository" "$install_directory"/data # Clone repo from Git
fi


FlightGear_Downloader.sh
#!/bin/bash

### ABOUT
# FlightGear_Downloader.sh
# This script will download, from Git repositories, the source code for
# the OpenSceneGraph, PLIB, SimGear and FlightGear.

### AUTHOR
# This software was created by Megaf - https://github.com/Megaf
# On the 20/10/2020

### LICENSE
########################################################################
#                                                                      #
#                       FREE AS IN FREEDOM LICENSE                     #
#                           FAIFL Version 0.2                          #
#                             October 2020                             #
#                                                                      #
#           Copyright (C) 2020 Megaf (https://github.com/Megaf/)       #
#                     <mmegaf [at] gmail [dot] com>                    #
#                                                                      #
# Everyone, anyone, anything or any company or organization, human,    #
# plant, apache helicopter, alive or not, form of life or not, is/are  #
# allowed to copy, modify, distribute, change, charge or do anything   #
# they/ones want with anything or whatever this license is, attached   #
# to or licensing.                                                     #
#                                                                      #
# This thing, software, hardware, piece, object, art, or whatever that #
# is been licensed under FAIFL (Free As In Freedom License) can be     #
# used, shared, copied, sold, modified, changed, adapted, or whatever  #
# you want to do as long as the following terms are understood and met.#
#                                                                      #
# (1) This that is licensed under FAIFL comes as it is.                #
# (2) This that is licensed under FAIFL it is not covered under any    #
# warranty.                                                            #
# (3) There is no warranty or guarantee that this that is licensed     #
# under FAIFL works at all.                                            #
# (4) This that is licensed under FAIFL may be dangerous and cause     #
# harm and loss of property.                                           #
# (5) Changing this license is not allowed.                            #
# (6) The creator, owner, author of this that is licensed under FAIFL  #
# shall not be liable,responsible or accused in any way for anything   #
# at all.                                                              #
#                                                                      #
########################################################################

# Using read command to get version information from the file VERSION
read -r version < VERSION

# Directory where the source files will be downloaded to.
download_directory="$HOME"/FlightGear-Stable_Source_Files

# Git Repository to download the source files from.
openscenegraph_repository="https://github.com/openscenegraph/OpenSceneGraph"
plib_repository="https://git.code.sf.net/p/libplib/code"
simgear_repository="git://git.code.sf.net/p/flightgear/simgear"
flightgear_repository="git://git.code.sf.net/p/flightgear/flightgear"

# Branches/Versions of each component.
openscenegraph_branch="OpenSceneGraph-3.6.5"
plib_branch="master"
#simgear_branch="version/2020.3.6"
#flightgear_branch="version/2020.3.6"
simgear_branch="release/2020.3"
flightgear_branch="release/2020.3"

# By default, without command line arguments the script will download the stable
# version of FlightGear, the following "if" statement downloads the "next"
# development version of FlightGear.
#
# It will also install FlightGear to folder name FlightGear-Next instead of 
# FlightGear-Stable
if [ "$*" = "--next" ]; then
    simgear_branch="next"
    flightgear_branch="next"
    download_directory="$HOME"/FlightGear-Next_Source_Files
fi

### CHECKS AND GIT CLONE
# We will perform a series of checks to see if the repos were cloned
# already or not, if they were not cloned yet, then git clone will run,
# Otherwise, the script will try to update, git pull, the repos.

### INTRODUCTION PART
clear
echo ""
echo "#====== Welcome to FlightGear Downloader $version !"
echo "#====== This script will download OpenSceneGraph, PLIB, SimGear and FlightGear."
echo ""
echo "#====== This script will not download nor update FlightGear Data. To do that,"
echo "#====== you can run the script \`./Download_Data.sh\` in a different terminal window."
echo "#====== While you run \`FlightGear_Downloader.sh\` and \`FlightGear_Compiler.sh\`"

# Waits for the user to press a key to continue.
echo ""
echo "======= Run this script with the flag --next to build the testing/next version of FlightGear."
echo "======= Selected FlighthGear version is $flightgear_branch"
echo ""
echo "#====== Press any key to continue or Ctrl + C to cancel."
read -rsn1
clear

# Checking OpenSceneGraph
echo ""
echo "#====== Checking if OpenSceneGraph was already downloaded."
if [ -d "$download_directory"/OSG ]; then
    echo "#======  OpenSceneGraph found, updating the repo if neded."
    cd "$download_directory"/OSG && git pull # Try to update existing previously cloned repo.
        else
            echo "#====== OpenSceneGraph not found, downloading now."
            git clone --depth=1 -b "$openscenegraph_branch" "$openscenegraph_repository" "$download_directory"/OSG # Clone repo from Git
fi

# Checking PLIB
echo ""
echo "#====== Checking if PLIB was already downloaded."
if [ -d "$download_directory"/PLIB ]; then
    echo "#======  PLIB found, updating the repo if neded."
    cd "$download_directory"/PLIB && git pull # Try to update existing previously cloned repo.
        else
            echo "#====== PLIB not found, downloading now."
            git clone --depth=1 -b "$plib_branch" "$plib_repository" "$download_directory"/PLIB # Clone repo from Git
fi

# Checking SimGear
echo ""
echo "#====== Checking if SimGear was already downloaded."
if [ -d "$download_directory"/SimGear ]; then
    echo "#======  SimGear found, updating the repo if neded."
    cd "$download_directory"/SimGear && git pull # Try to update existing previously cloned repo.
        else
            echo "#====== SimGear not found, downloading now."
            git clone -b "$simgear_branch" "$simgear_repository" "$download_directory"/SimGear # Clone repo from Git
fi

# Checking FlightGear
echo ""
echo "#====== Checking if FlightGear was already downloaded."
if [ -d "$download_directory"/FlightGear ]; then
    echo "#======  FlightGear found, updating the repo if neded."
    cd "$download_directory"/FlightGear && git pull # Try to update existing previously cloned repo.
        else
            echo "#====== FlightGear not found, downloading now."
            git clone -b "$flightgear_branch" "$flightgear_repository" "$download_directory"/FlightGear # Clone repo from Git
fi

echo ""
echo "#====== If everything went well then you can compile FlightGear"
echo "#====== by running the command ./FlightGear_Compiler.sh"
echo "#====== Remember, you can run Download_Data.sh at the same time you"
echo "#====== run any of the other scripts."
echo ""
echo "#====== Feel free to ask for help and contribute to this script at"
echo "#====== https://github.com/Megaf/FlightGearBuilder "
echo ""


FlightGearBuilder.sh
#!/bin/bash

# First function of this shell script.
# This fuction runs sudo apt to install all dependencies that are required for a complete OSG, SG and FG build. PLIB is listed as a dependency here
# and will not be compiled, but installed from the distros repository.
# This function at the present time is only for Debian based distros.
deb_install_deps() {
    clear
    echo "======================================"
    echo "=> Hello there."
    echo "=> I hope you are having a good time!"
    echo "=> If not, I wish you all the best!"
    echo "--------------------------------------"
    echo "=> - Megaf mmegaf [at] gmail [dot] com"
    echo "======================================"
    echo ""
    echo "=> We are about to install all the dependencies required to compile FlightGear."
    echo "-------------------------------------------------------------------------------"
    echo ""
    echo "=> You will be asked, by sudo, to enter your user password so it can use apt to"
    echo "=> install all the required libs and software."
    echo ""
    echo "=> If you want to skip this step, press Ctrl + c."
    echo ""
    sudo apt install -y subversion build-essential git cmake ccache libgmp-dev subversion libboost-dev freeglut3-dev libopenal-dev zlib1g-dev liblzma-dev libcurl4-gnutls-dev libsdl2-dev libsdl1.2-dev libpoppler-glib-dev libjsoncpp-dev libudev-dev libnvtt-dev libnvtt2 libfreetype6-dev libsdl2-dev libsdl2-gfx-dev libsdl2-image-dev libsdl2-mixer-dev libsdl2-net-dev libsdl2-ttf-dev libjpeg-dev libxml2-dev libtiff-dev libpng-dev qtdeclarative5-dev qttools5-dev qtbase5-dev-tools qttools5-dev-tools qml-module-qtquick2 qml-module-qtquick-window2 qml-module-qtquick-dialogs libqt5opengl5-dev libqt5svg5-dev libqt5websockets5-dev qml-module-qtquick-controls2 libgtk2.0-dev librsvg2-dev libgtkglext1-dev libgtkglextmm-x11-1.2-dev libevent-dev libevent-extra-2.1-6 libevent-pthreads-2.1-6 libglew-dev # libplib-dev qt5-default
}

# Here we are setting some variables that will be used for the whole script.

# Reading install destination for the file DESTINATION. Edit that file with the full path of where you want FlightGear to be installed to.
# If you want to install to a USB drive called USDdrive, add "/media/$USER/USDdrive" to DESTINATION.
instroot="$(envsubst < DESTINATION)"

# In this first batch of variables we are setting for the stable and default version of FlightGear. ie, without using --next
fg_install="$instroot""/FGB/FlightGear-Stable" # Where FlightGear will be installed to.
fg_branch="release/2020.3" # Defining the branch we want to compile against, at the moment Im writing this the latest stable is 2020.3.
osg_branch="OpenSceneGraph-3.6" # Defining branch and version for OpenSceneGraph.
dir="$instroot""/FGB/FlightGear-Sources-Stable" # Where the source code for OSG, SG and FG will be kept. A full download is done first. Later updated.
ldlib="export LD_LIBRARY_PATH=""$fg_install""/lib:""$fg_install""/lib64" # All our libs are located withing the FG install location.
release_type="Stable" # Will be used to set the launcher name in the desktop and menus.
CFLAGS="-w -march=native -mtune=native -Ofast"
CXXFLAGS="$CFLAGS"
cc_flags="$CFLAGS"

# This batch of variables are similar to the above, here we simply check if the user wants to build FG Next and adjust variables accordinly.
if [ "$*" = "--next" ]; then
    fg_install="$instroot""/FGB/FlightGear-Next"
    fg_branch="next"
#    osg_branch="master" # Uncomment this line to use the latest development version of OSG for FlightGear Next.
    dir="$instroot""/FGB/FlightGear-Sources-Next"
    ldlib="export LD_LIBRARY_PATH=""$fg_install""/lib"
    release_type="Next"
fi

# Our second function
# Some general variables, defining locations of files and directories for FlightGear.
set_globals() {
    fg_common="$instroot""/FGB/FlightGear-Common" # FlightGear-Common is where we put FG_HOME and downloaded content.
    fg_home="$fg_common/FG_HOME" # FG_HOME is where logs and settings live.
    fg_scenery="$fg_common/Scenery" # Scenery will be downloaded by terrasync to Scenery, inside FlightGear-Common. Add custom scenery here.
    fg_aircraft="$fg_common/Aircraft" # Add your downloaded aircraft here, to Aircraft in FlightGear-Common
    export PATH="/usr/lib/ccache:$PATH" # This is so the compiler actually caches what's built, so it can reuse compile code when possible.
    echo "=> Here are our directories."
    echo ""
    echo "=> The downloaded source codes are located at $dir"
    echo "=> FG_HOME, Scenery and Aircraft are located at $fg_common"
    echo "=> Your FG_HOME, the place where the settings and logs live is $fg_home"
    echo "=> TerraSunk will download the scenery to $fg_scenery, you can add custom scenery there."
    echo "=> $fg_aircraft Is where you can put your manually downloaded aircraft."
}

# Third function of our script, how the source will be downloaded
svn_checkout() {
    echo ""
    echo "=> Now downloading $target from $url"
    local to_check="$dir""/$target"
    if [ -d "$to_check" ]; then
        cd $to_check && svn update
    else
        svn checkout "$url"/"$target" "$to_check"
    fi
}

git_clone() {
    echo ""
    echo "=> Now downloading $target from $url"
    local to_check="$dir""/$target"
    if [ -d "$to_check" ]; then
        cd $to_check && git checkout "$branch" && git pull
    else
        git clone -b "$branch" "$url"/"$target" "$to_check"
    fi
}

compile() {
    echo "oo Compiling $target oo"
    echo "dir = ""$dir"
    echo "target = ""$target"
    cd "$dir""/""$target"
    ./autogen.sh
    ./configure --prefix="$fg_install"
    make clean && make -j "$(expr $(nproc) '+' 2)"
}

build() {
    echo "oo Compiling $target oo"
    local cmake_flags="-DCMAKE_C_FLAGS=$cc_flags -DCMAKE_CXX_FLAGS=$cc_flags -DCMAKE_INSTALL_PREFIX=$fg_install -DCMAKE_BUILD_TYPE=Release $cmflags"
    local bld_dir="$dir"/building/"$target"
    rm -rf "bld_dir" && mkdir -p "$bld_dir" && cd "$bld_dir"
    cmake $cmake_flags "$dir"/"$target"
    make clean && make -j "$(expr $(nproc) '+' 2)"
}

make_runner() {
    echo "oo Generating Runner oo"
cat << EOF > "$fg_install"/flightgear
#!/bin/bash
export OSG_NUM_DATABASE_THREADS="$(expr $(nproc) '+' 1)"
export OSG_OPTIMIZER=REMOVE_REDUNDANT_NODES
export OSG_NUM_HTTP_DATABASE_THREADS="$(expr $(nproc) '+' 1)"
export OSG_GL_TEXTURE_STORAGE="on"
export OSG_COMPILE_CONTEXTS="on"
export FG_HOME="$fg_home"
export FG_PROG="$fg_install"
export FG_SCENERY="$fg_scenery"
$ldlib
"$fg_install"/bin/fgfs --fg-aircraft="$fg_aircraft" --terrasync-dir="$fg_scenery" --prop:/sim/rendering/multithreading-mode=CullThreadPerCameraDrawThreadPerContext --prop:/sim/gui/current-style=0 --prop:/sim/nasal-gc-threaded=true --prop:/sim/rendering/use-vbos=true \$*
EOF

chmod +x "$fg_install"/flightgear

cat << EOF > "$fg_install"/launcher_loop
#!/bin/bash

while :
do
    "$fg_install"/flightgear --enable-fullscreen --launcher
done
EOF

chmod +x "$fg_install"/launcher_loop
}

install() {
    make install
}

set_globals

# Install dependencies
deb_install_deps

# Build and Install PLIB
url="svn://svn.code.sf.net/p/plib/code"
target="trunk"
svn_checkout
compile
install
unset target url

# Build and Install OpenSceneGraph
url="https://github.com/openscenegraph"
target="OpenSceneGraph.git"
branch="$osg_branch"
cmflags="-DBUILD_OSG_APPLICATIONS=0 -DBUILD_OSG_DEPRECATED_SERIALIZERS=0 -DOpenGL_GL_PREFERENCE=GLVND"
git_clone
build
install
unset cmflags target url branch

# Build and Install SimGear
url="https://gitlab.com/flightgear"
target="simgear.git"
branch="$fg_branch"
cmflags="-DBUILD_TESTING=0 -DENABLE_SIMD_CODE=1 -DENABLE_TESTS=0"
git_clone
build
install
unset cmflags target url branch

# Build and Install FlightGear
url="https://gitlab.com/flightgear"
target="flightgear.git"
branch="$fg_branch"
cmflags="-DFG_BUILD_TYPE=Release -DBUILD_TESTING=0 -DENABLE_AUTOTESTING=0"
git_clone
build
install
unset cmflags target url branch

# Making the launcher
make_runner
mkdir -p "$fg_common"
mkdir -p "$fg_home" "$fg_scenery" "$fg_aircraft"

echo ""
echo "#====== Creating a Desktop shortcut for FlightGear-$release_type"

release="$fg_branch"
install_directory="$fg_install"

# Creates desktop luncher
cat << EOF > "$HOME""/Desktop/FlightGear-$release_type.desktop"
#!/usr/bin/env xdg-open
[Desktop Entry]
Version=1.4
Type=Application
Categories=Game;Simulation
Terminal=false
Icon=$install_directory/share/icons/hicolor/scalable/apps/flightgear.svg
Path=$install_directory
#Exec=sh -c "cd $install_directory && ./flightgear --launcher"
Exec=./$install_directory/flightgear --launcher
Name=FlightGear-$release_type
Comment=FlightGear-$release_type Launcher Compiled with FlightGear Builder
Hidden=false
Keywords=Flight Simulator;Simulation;Flight;FlightGear;FlightGear Builder;FGB;Aviation;Airplanes
EOF
chmod +x "$HOME""/Desktop/FlightGear-""$release_type"".desktop" # Sets it as executable

echo ""
echo "#====== Adding FlightGear-$release_type menu entry."

mkdir -p "$HOME"/.local/share/applications # Creating this directory if it doesn't exist
cp "$HOME"/Desktop/FlightGear-"$release_type".desktop "$HOME"/.local/share/applications/ # Adds FlightGear to the list of software in the users menu.
